jdk-http-client/src/main/scala/org/http4s/client/jdkhttpclient/JdkHttpClient.scala

```diff
@@ -0,0 +1,131 @@
+import org.http4s.internal.fromCompletableFuture
+import org.http4s.util.CaseInsensitiveString
+import org.http4s.{Header, Headers, HttpVersion, Request, Response, Status}
+import org.reactivestreams.FlowAdapters
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+
+// TODO documentation
+object JdkHttpClient {
+
+  def apply[F[_]](
+      jdkHttpClient: HttpClient,
+      ignoredHeaders: Set[CaseInsensitiveString] = restrictedHeaders
+  )(implicit F: ConcurrentEffect[F]): Client[F] = {
+
+    def convertRequest(req: Request[F]): F[HttpRequest] =
+      convertHttpVersionFromHttp4s[F](req.httpVersion).map { version =>
+        val rb = HttpRequest.newBuilder
+          .method(
+            req.method.name, {
+              val publisher = FlowAdapters.toFlowPublisher(
+                StreamUnicastPublisher(req.body.chunks.map(_.toByteBuffer)))
+              if (req.isChunked)
+                BodyPublishers.fromPublisher(publisher)
+              else
+                req.contentLength.fold(BodyPublishers.noBody)(
+                  BodyPublishers.fromPublisher(publisher, _))
+            }
+          )
+          .uri(URI.create(req.uri.renderString))
+          .version(version)
+        val headers = req.headers.iterator
+        // hacky workaround (see e.g. https://stackoverflow.com/questions/53979173)
+          .filterNot(h => ignoredHeaders.contains(h.name))
+          .flatMap(h => Iterator(h.name.value, h.value))
+          .toArray
+        (if (headers.isEmpty) rb else rb.headers(headers: _*)).build
+      }
+
+    def convertResponse(res: HttpResponse[Flow.Publisher[util.List[ByteBuffer]]]): F[Response[F]] =
+      F.fromEither(Status.fromInt(res.statusCode)).map { status =>
+        Response(
+          status = status,
+          headers = Headers(res.headers.map.asScala.flatMap {
+            case (k, vs) => vs.asScala.map(Header(k, _))
+          }.toList),
+          httpVersion = res.version match {
+            case HttpClient.Version.HTTP_1_1 => HttpVersion.`HTTP/1.1`
+            case HttpClient.Version.HTTP_2 => HttpVersion.`HTTP/2.0`
+          },
+          body = FlowAdapters
+            .toPublisher(res.body)
+            .toStream[F]
+            .flatMap(bs =>
+              Stream.fromIterator(bs.asScala.map(Chunk.byteBuffer).iterator).flatMap(Stream.chunk))
+        )
+      }
+
+    Client[F] { req =>
+      Resource.liftF(
+        convertRequest(req)
+          .flatMap(r =>
+            fromCompletableFuture(F.delay(jdkHttpClient.sendAsync(r, BodyHandlers.ofPublisher))))
+          .flatMap(convertResponse))
+    }
+  }
+
+  def createUnderlyingClient[F[_]](
+      httpVersion: HttpVersion = HttpVersion.`HTTP/1.1`,
+      connectTimeout: Duration = 10.seconds,
+      redirectPolicy: HttpClient.Redirect = HttpClient.Redirect.NEVER,
+      sslContext: => SSLContext = SSLContext.getDefault,
+      proxySelector: => ProxySelector = ProxySelector.getDefault,
```

Comment:
There might be better names for this function. `sslContext` and `proxySelector` are by-name, as they can throw (very unlikely) - not the best solution.

Reply:
We have seen `SSLContext` throw before in blaze.

I'm not so sure this function carries its weight over just using their builder.  I guess we get `HttpClient.Version` parsing out of it?

Reply:
What about removing `sslContext` and `proxySelector`? Even without them, I think most uses are coverd by this.

Reply:
I've never used `ProxySelector`.  `SSLContext` is something I very often customize (typically for mutual auth), but we can always use the builder.

Maybe the function to convert our HttpVersion to their version should be public?  It's probably the most irritating part of this constructor to reinvent if someone does need a custom context or selector.

Reply:
Ok, so remove `createUnderlyingClient` and make `convertHttpVersionFromHttp4s` public? What about `simple`? I am in favor of keeping it, as it is non-obvious that `HttpClient.newBuilder().build()` (or equivalently, `HttpClient.newHttpClient()`) is side-effectful. What about this:
```scala
def simple[F[_]](implicit F: ConcurrentEffect[F]): F[Client[F]] =
  F.delay(HttpClient.newHttpClient()).map(apply(_))
```